### PR TITLE
Remove Bool/IndexTensor from schema for native functions with derivatives

### PR DESCRIPTION
**Pull request by cpuhrsch**
This only deals with four functions, but is an important first step towards removing BoolTensor and IndexTensor entirely.

### CODE REVIEW

File: tools/autograd/gen_variable_type.py [hunk 1]
```diff
@@ -485,11 +485,32 @@ def is_differentiable(arg):
         if 'Tensor' not in arg['type']:
             return False
         if arg['dynamic_type'] in {'IndexTensor', 'BoolTensor'}:
+            # TODO: Enable this after native_functions.yaml schema unification.
+            # These are necessary for legacy code and should be
+            # used by legacy code only!
+            # assert name.startswith('_th_'), \
```
Comment (gchanan) [inline tools/autograd/gen_variable_type.py]: isn't there a mode or something you can check?

File: tools/autograd/gen_variable_type.py [hunk 2]
```diff
@@ -485,11 +485,32 @@ def is_differentiable(arg):
...
+            raise RuntimeError('Missing arguments for derivatives: {} in {}'.format(missing, func['name']))
+        return differentiable
+
     inputs = [arg for arg in arguments if not arg.get('output', False)]
     differentiable_inputs = list(filter(is_differentiable, inputs))
+    args_with_derivatives = find_args_with_derivatives(differentiable_inputs)
+    not_args_with_derivatives = []
+    if func:
+        for not_derivative in func['not_derivatives']:
+            not_args_with_derivatives += not_derivative['var_names']
```
Comment (gchanan) [inline tools/autograd/gen_variable_type.py]: append?

File: tools/autograd/load_derivatives.py [hunk 3]
```diff
@@ -144,15 +144,25 @@ def set_up_derivatives(defn_name, defn, declaration):
...
         # Set up the derivative information
         derivatives = []
+        not_derivatives = []
+        not_args_with_gradients = []
         for raw_names in sorted(defn.keys()):
             formula = defn[raw_names]
             names = split_names(raw_names)
-            derivatives.append(create_derivative(declaration, formula, names))
+            if formula.lower().strip() == 'not_differentiable':
+                derivative = create_derivative(declaration['arguments'], declaration['returns'],
```
Comment (gchanan) [inline tools/autograd/load_derivatives.py]: since you unconditionally create the derivative can you not put it in an if / else.

File: tools/autograd/load_derivatives.py [hunk 4]
```diff
@@ -144,15 +144,25 @@ def set_up_derivatives(defn_name, defn, declaration):
...
+                                               declaration['name'], formula, names)
+                not_derivatives.append(derivative)
+                not_args_with_gradients += derivative['var_names']
```
Comment (gchanan) [inline tools/autograd/load_derivatives.py]: append.

File: tools/autograd/derivatives.yaml [hunk 5]
```diff
@@ -8,6 +8,9 @@
 #     Note that a single gradient entry can specify the gradient
 #     formula for multiple input names, by specifying a key
 #     "input1, input2" (see atan2 for an example).
+#   - An argument can be flagged as 'not_differentiable'. That means this
+#     argument cannot have a derivative defined and must not be used in any
+#     derivative calculations.
```
Comment (gchanan) [inline tools/autograd/derivatives.yaml]: This needs an explanation for how you handle "not_differentiable" vs not specifying the derivative.

File: tools/autograd/load_derivatives.py [hunk 6]
```diff
@@ -144,15 +144,25 @@ def set_up_derivatives(defn_name, defn, declaration):
 
         # Set up the derivative information
         derivatives = []
+        not_derivatives = []
```
Comment (gchanan) [inline tools/autograd/load_derivatives.py]: I don't know what these names mean.

**Review by gchanan — COMMENTED** [review]
can you post the generated code differences?

File: tools/autograd/load_derivatives.py [hunk 7]
```diff
@@ -144,15 +144,25 @@ def set_up_derivatives(defn_name, defn, declaration):
...
-            derivatives.append(create_derivative(declaration, formula, names))
+            if formula.lower().strip() == 'not_differentiable':
+                derivative = create_derivative(declaration['arguments'], declaration['returns'],
+                                               declaration['name'], formula, names)
+                not_derivatives.append(derivative)
+                not_args_with_gradients += derivative['var_names']
+            else:
+                derivatives.append(create_derivative(declaration['arguments'], declaration['returns'],
+                                                     declaration['name'], formula, names))
+        args_with_gradients = list(filter(lambda x: x['name'] not in not_args_with_gradients, args_with_gradients))
```
Comment (gchanan) [inline tools/autograd/load_derivatives.py]: does this have any effect?  What is this supposed to catch?  If all the arguments don't have gradients?

File: tools/autograd/gen_variable_type.py (hunk 1 above)
Comment (cpuhrsch) [reply tools/autograd/gen_variable_type.py]: Maybe, I'll revisit this once IndexTensor/BoolTensor has been entirely removed. Until then this needs to exist as a general reminder.

**Comment by cpuhrsch** [comment]
[Here](https://gist.github.com/cpuhrsch/2af382ffe9635df996dcfcef57a0b371) is the changeset for the generated code.

File: tools/autograd/load_derivatives.py (hunk 4 above)
Comment (cpuhrsch) [reply tools/autograd/load_derivatives.py]: What's the advantage? A formula can have multiple args and we need to flatten them into a single list.

File: tools/autograd/gen_variable_type.py (hunk 2 above)
Comment (cpuhrsch) [reply tools/autograd/gen_variable_type.py]: No, they need to be flattened. A formula can refer to multiple args.

File: tools/autograd/load_derivatives.py (hunk 6 above)
Comment (cpuhrsch) [reply tools/autograd/load_derivatives.py]: I'll choose a better name.

File: tools/autograd/gen_variable_type.py (hunk 1 above)
Comment (gchanan) [reply tools/autograd/gen_variable_type.py]: sound good.  I think checking that the mode isn't 'TH' and isn't 'NN" is sufficient.

File: tools/autograd/gen_variable_type.py (hunk 2 above)
Comment (gchanan) [reply tools/autograd/gen_variable_type.py]: ah, my bad.  There are no cases where this is actually used for not_derivatives right?

**Comment by gchanan** [comment]
generated code changes look good.

File: tools/autograd/load_derivatives.py [hunk 8]
```diff
@@ -144,15 +144,25 @@ def set_up_derivatives(defn_name, defn, declaration):
...
+            derivative = create_derivative(declaration['arguments'], declaration['returns'],
+                                           declaration['name'], formula, names)
+            if formula.lower().strip() == 'not_differentiable':
+                assert not sum([type(var_name) == list
+                                for var_name in derivative['var_names']]), \
+                    "Variable names associated to a formula should be a flat list"
+                args_with_no_gradients += derivative['var_names']
+            else:
+                derivatives.append(derivative)
+        args_with_gradients = list(filter(lambda x: x['name'] not in args_with_no_gradients, args_with_gradients))
```
Comment (gchanan) [inline tools/autograd/load_derivatives.py]: I still don't know what this is supposed to be doing -- it seems like if something is in both args_with_gradients and args_with_no_gradients it should be error, not that it should silently removed from args_with_gradients.

File: tools/autograd/load_derivatives.py [hunk 9]
```diff
@@ -144,15 +144,25 @@ def set_up_derivatives(defn_name, defn, declaration):
 
         # Set up the derivative information
         derivatives = []
+        args_with_no_gradients = []
```
Comment (gchanan) [inline tools/autograd/load_derivatives.py]: this name is still strange because there are 3 possibilities:
1) an arg has an entry with a specified gradient
2) an arg has an entry specified as not differentiable
3) an arg has no entry

I would expect `args_with_no_gradients` applies to 3), but it actually applies to 2) and this isn't explained.

File: tools/autograd/gen_variable_type.py [hunk 10]
```diff
@@ -598,24 +615,15 @@ def setup_derivative():
...
-        return differentiable
-
     def emit_check_no_requires_grad(tensor_args, args_with_derivatives):
         """Checks that arguments without derivatives don't require grad"""
         body = []
         for arg in tensor_args:
             if arg in args_with_derivatives:
                 continue
             name = arg['name']
+            if name in args_with_no_gradients:
```
Comment (gchanan) [inline tools/autograd/gen_variable_type.py]: this is super confusing because of the non-congruence of args_with_derivatives (full arguments) and args_with_no-gradients (just names).  I'm not sure why this doesn't use args_with_gradients, but I guess this was an issue before your change...

Comment (cpuhrsch) [reply tools/autograd/gen_variable_type.py]: Yes, variable_type uses derivatives, whereas load_derivatives uses gradients. I'll just change gradients to derivatives.

File: tools/autograd/load_derivatives.py (hunk 9 above)
Comment (cpuhrsch) [reply tools/autograd/load_derivatives.py]: Fair point, let's go for 'not_differentiable_args'.

File: tools/autograd/derivatives.yaml [hunk 11]
```diff
@@ -8,6 +8,15 @@
...
 #     "input1, input2" (see atan2 for an example).
+#   - An argument can be flagged as 'not_differentiable'.
+#     In general there are 3 possibilities:
+#       1. An argument has an entry with a specified gradient
+#       2. An argument has an entry specified as not differentiable
+#       3. An argument has no entry
+#     Using the flag 'not_differentiable' resolves to the second case.
+#     The second case was introduced in support for arguments of
+#     type e.g. IndexTensor for 'embedding', that are not differentiable.
+#     TODO: Determine whether case 3 and case 2 can be replaced by one concept.
```
Comment (gchanan) [inline tools/autograd/derivatives.yaml]: it would also be nice if you explained what the behavior of 3 is currently.

File: tools/autograd/gen_variable_type.py [hunk 12]
```diff
@@ -483,11 +483,29 @@ def is_differentiable(arg):
...
+        names = set(name for d in func['derivatives'] for name in d['var_names'])
+        differentiable = [arg for arg in differentiable_inputs if arg['name'] in names]
+        if len(differentiable) != len(names):
+            missing = names - set(arg['name'] for arg in differentiable)
+            raise RuntimeError('Missing arguments for derivatives: {} in {}'.format(missing, func['name']))
+        return differentiable
+
     inputs = [arg for arg in arguments if not arg.get('output', False)]
     differentiable_inputs = list(filter(is_differentiable, inputs))
+    args_with_derivatives = find_args_with_derivatives(differentiable_inputs)
```
Comment (gchanan) [inline tools/autograd/gen_variable_type.py]: I think this renaming is actually worse.  The previous naming IMO had 2 problems:
1) one thing was called gradients and one was called derivatives and it's not clear what difference was implied and/or why there wasn't just one concept.
2) both `args_with_derivatives` and `not_differentiable_args` are both called args, but for one it's dict of the entire arg and the other it's just a name.

This renaming doesn't fix 2) and it hides the difference in 1), i.e. there are still two concepts but now they have the same name.

**Review by gchanan — APPROVED** [review]

File: tools/autograd/gen_variable_type.py (hunk 12 above)
Comment (cpuhrsch) [reply tools/autograd/gen_variable_type.py]: On 1) I don't think there was a difference between the two. The names were used interchangeably. 

On 2), i'll add "_names" to "not_differentiable_args".

File: tools/autograd/derivatives.yaml (hunk 11 above)
Comment (cpuhrsch) [reply tools/autograd/derivatives.yaml]: That is indeed not entirely clearly defined. See the TODO. This is the kind of stuff we set out to discover with this kind of work.
